telemetry: Add `language_name` and `model_provider`

bcherny opened the pull request:
This PR adds a bit more metadata for assistant logging.

Release Notes:

- Assistant: Added `language_name` and `model_provider` fields to telemetry events.

bcherny committed:
telemetry: add language_name to AssistantEvent

bcherny committed:
telemetry: add model_provider to AssistantEvent

bcherny commented:
@cla-bot check

bcherny commented on crates/assistant/src/context.rs:
```diff
@@ -2133,14 +2133,17 @@ impl Context {
                     });
 
                     if let Some(telemetry) = this.telemetry.as_ref() {
-                        telemetry.report_assistant_event(
```
Ran into the Clippy 7-arg limit when I just added to this

maxdeviant committed:
Merge branch 'main' into pr/18640

maxdeviant committed:
Don't expose internal `report_event` method

rgbkrk commented on crates/client/src/telemetry.rs:
```diff
@@ -536,7 +541,7 @@ impl Telemetry {
     }
 
-    pub fn report_event(self: &Arc<Self>, event: Event) {
+    fn report_event(self: &Arc<Self>, event: Event) {
```
👍

maxdeviant committed:
Merge branch 'main' into pr/18640

maxdeviant committed:
Expand variable name

maxdeviant commented on crates/multi_buffer/src/multi_buffer.rs:
```diff
@@ -382,6 +382,15 @@ impl MultiBuffer {
     }
 
+    pub fn language<'a>(&self, cx: &'a AppContext) -> Option<&'a Arc<Language>> {
+        // Take the last language, since all buffer are guaranteed to be
```
I don't think this is true?

What is guaranteeing that each buffer in the multibuffer are in the same file?

bcherny replied:
I don't believe there is a way to open an inline assistant across files, right? All the buffers must be in the same file from what I could tell.

maxdeviant replied:
Even if it were, that property is not true for the multibuffer as a whole, so having this `language` getter available for every multibuffer is not correct.

We'd need to do something that is specific to the inline assist case.

> I don't believe there is a way to open an inline assistant across files, right?

It is possible to have multiple inline assists open in a multibuffer that spans multiple files. Here's an example where I initiated an inline assist from within a project search:

<img width="868" alt="Screenshot 2024-10-04 at 1 38 55 PM" src="https://github.com/user-attachments/assets/366149ed-8e44-4949-886f-b17a7b59c483">

bcherny replied:
Gotcha. Any suggestions for how to grab the correct buffer for a given inline assistant? If that's hard to do, we can also send up a Vec of languagenames.

maxdeviant replied:
Max and I are taking a look at it now.

maxdeviant replied:
We removed the `language` method in the multibuffer and addressed inferring the language for the assist on a case-by-case basis.

We're able to use the range of the inline assistant to work out which buffer we're in and get that language's multibuffer:

https://github.com/zed-industries/zed/pull/18640/commits/12b95cedcda5be542e547c6b3079a863f505241f

bcherny replied:
Oh that looks solid! Thank you for jumping in.

maxdeviant reviewed (changes requested):
How we're determining the language for a multibuffer doesn't seem correct.

maxdeviant committed:
Remove unneeded `to_string`

maxdeviant committed:
Inline `language` method

We'll need to figure out the best way to get the language on a case-by-case basis

maxdeviant committed:
Update how we retrieve the language for the inline assist events

Co-authored-by: Max <max@zed.dev>

maxdeviant committed:
Use `AssistantKind` already in scope

maxdeviant reviewed (approved):
Thanks!
